heltour/tournament/tests/test_admin.py

```diff
@@ -1,50 +1,501 @@
+        )
+        cls.p1 = Player.objects.get(lichess_username="Player1")
+        cls.p2 = Player.objects.get(lichess_username="Player2")
+        cls.p3 = Player.objects.get(lichess_username="Player3")
+        cls.p4 = Player.objects.get(lichess_username="Player4")
+        cls.s = get_season("team")
+        cls.sp1 = SeasonPlayer.objects.create(player=cls.p1, season=cls.s)
+        cls.path_s_changelist = reverse("admin:tournament_season_changelist")
+        cls.path_m_p = reverse("admin:manage_players", args=[cls.s.pk])
+
+    @patch("heltour.tournament.simulation.simulate_season")
+    @patch("django.contrib.admin.ModelAdmin.message_user")
+    def test_simulate(self, message, simulate):
+        self.client.force_login(user=self.superuser)
+        self.client.post(
+            self.path_s_changelist,
+            data={
+                "action": "simulate_tournament",
+                "_selected_action": get_season("lone").pk,
+            },
+            follow=True,
+        )
+        self.assertTrue(simulate.called)
+        self.assertTrue(message.called)
+        self.assertEqual(message.call_args.args[1], "Simulation complete.")
+
+    @patch("heltour.tournament.models.Season.calculate_scores")
+    @patch("heltour.tournament.models.TeamPairing.refresh_points")
+    @patch("heltour.tournament.models.TeamPairing.save")
+    @patch("django.contrib.admin.ModelAdmin.message_user")
+    def test_recalculate(self, message, tpsave, tprefresh, scalculate):
+        self.client.force_login(user=self.superuser)
+        self.client.post(
+            self.path_s_changelist,
+            data={
+                "action": "recalculate_scores",
+                "_selected_action": get_season("lone").pk,
+            },
+            follow=True,
+        )
+        self.assertFalse(tprefresh.called)
+        self.assertFalse(tpsave.called)
+        self.assertTrue(scalculate.called)
+        self.assertTrue(message.called)
+        self.assertEqual(message.call_args.args[1], "Scores recalculated.")
+        message.reset_mock()
+        scalculate.reset_mock()
+        self.client.post(
+            self.path_s_changelist,
+            data={
+                "action": "recalculate_scores",
+                "_selected_action": Season.objects.all().values_list("pk", flat=True),
+            },
+            follow=True,
+        )
+        self.assertTrue(tprefresh.called)
+        self.assertTrue(tpsave.called)
+        self.assertTrue(scalculate.called)
+        self.assertEqual(scalculate.call_count, 2)
+        self.assertTrue(message.called)
+        self.assertEqual(message.call_args.args[1], "Scores recalculated.")
+
+    @patch("django.contrib.admin.ModelAdmin.message_user")
+    @patch(
+        "heltour.tournament.admin.normalize_gamelink",
+        side_effect=[("incorrectlink1", False), ("mockedlink2", True)],
+    )
+    def test_verify(self, gamelink, message):
```

Comment:
Yeah, I'm confused by the order of these args, are they really in reverse order?